HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,90 @@
+# Initialize Spark
+from pyspark import SparkContext
+sc = SparkContext()
+sc.setLogLevel("ERROR")
+
+# Helper functions for cleaning wiki data
+def link_string_to_KV(s):
+    src, dests = s.split(': ')
+    dests = [int(to) for to in dests.split(' ')]
+    return (int(src), dests)
+
+# Helper function for updating path
+def get_path(x):
+    newChars = x[1][1] 
+    oldPath = x[1][0][1]
+    return [(c, [len(oldPath + [c]), oldPath + [c]]) for c in newChars]
+
+# Finds the shortest path between two nodes
+def find_path(searchGraph, startNodeID, endNodeID):
+    
+    # Initialize key variables
+    searchNodes = sc.parallelize([startNodeID]).map(lambda x: (x, x))
+    explored = sc.parallelize([startNodeID]).map(lambda x: (x, x))
+    frontier = sc.parallelize([startNodeID]).map(lambda x: (x, [1, [x]]))
+    targetNodes = sc.parallelize([])
+    dist = 1
+    
+    # Keep looping until the target node has been found
+    while targetNodes.isEmpty():
+
+        dist += 1
+        
+        # Look up new children nodes
+        children = searchGraph.join(searchNodes).map(lambda x: (x[0], x[1][0])).partitionBy(256)
+        
+        # Update the paths
+        frontier = frontier.join(children).map(lambda x: get_path(x)).flatMap(lambda x: x).partitionBy(256)
+        
+        # Look up all paths that end in the target node
+        targetNodes = frontier.filter(lambda x: x[0]==endNodeID).cache()
+        
+        # Update the list of nodes to search in the next iteration
+        searchNodes = frontier.filter(lambda x: x[1][0]==dist).map(lambda x: (x[0], x[0])).distinct().subtractByKey(explored).partitionBy(256).cache()
+
+        # Update the list of explored nodes
+        explored = frontier.flatMap(lambda x: x[1][1]).map(lambda x: (x, x)).distinct().partitionBy(256).cache()
+        
+    return targetNodes.map(lambda x: x[1][1]).collect()
```

Comment:
It's actually just one collect after all the iterations are completed (the indentation may not be showing up correctly on your end?).  I wanted to print out all the possible paths in order to compare the number of paths from KB > HU vs HU > KB, so I figured that one collect at the end was inevitable.